src/Microsoft.AspNet.Routing/RouteData.cs

```diff
@@ -32,24 +34,55 @@ public RouteData(RouteData other)
                 throw new ArgumentNullException(nameof(other));
             }
 
-            DataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
             Routers = new List<IRouter>(other.Routers);
-            Values = new RouteValueDictionary(other.Values);
+            
+            // Perf: Avoid allocating DataTokens and RouteValues unless we need to make a copy.
+            if (other._dataTokens != null)
+            {
+                _dataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
```

Comment:
Another options I considered as well is to add a `_parent` field and further defer that copy. I guess I'll try that also and see if it's better.